readme.md
# sketch-ui

__Warning: extremely early version. Breaking changes will happen.__

Generate a UI for tweaking the variables in your creative coding sketches with minimal configuration, using standard HTML inputs.

Designed for use with [p5.js](https://p5js.org/) but not bound to any library or framework.

## Installation
The examples currently assume you're using Webpack to build your JS and CSS.

Add it to your project:
```shell
npm install @velvetkevorkian/sketch-ui
```

## Usage

```javascript
import UI from '@velvetkevorkian/sketch-ui'
import '@velvetkevorkian/sketch-ui/src/ui.css'

// declare our variables object.
const vars = {
  myVariable: {
    value: 50
  }
}

// create a new UI instance and pass the variables object to it
const ui = new UI(vars).proxy

function loop() {
  // vars.myVariable.value is proxied to ui.myVariable
  console.log(ui.myVariable)
  // keep the loop going
  window.requestAnimationFrame(loop)
}

// start the requestAnimationFrame loop
window.requestAnimationFrame(loop)
```

Drag the slider and watch the values update in real time.

Setting a value works the same way.
```javascript
  ui.myVariable = 200
  console.log(ui.myVariable) // 200
```

By default a number using a `range` input has a range of 0-255. Trying to set a variable outside that range will see it clamped to 0 or 255. Let's look at how we can configure the options.

## Configuration
The script will try to infer the type of variable you're using, and generate a suitable input with sensible defaults:
- 7 character hex codes, e.g. `#ff0000`, generates a `color` input.
- numbers generates a `range` input with `min=0`, `max=255`, and `step=1`
- a boolean value generates a `checkbox` input
- an array generates a `select` field with each item in the array an `option`
- anything else will default to a standard `text` input

You can pass additional options to override any of these defaults. For example:
```javascript

const vars = {
  // use a `number` input instead of `range`
  num: {
    value: 50,
    type: 'number'
  },
  // override the step and max values
  bigNum: {
    value: 100,
    max: 1000,
    step: 10
  }
}
```
The script will validate `min` and `max` values for setters as well as via the UI, which is why `myVariable` maxes out at 255 in the click example above.

You can also set a custom label value if you don't want to use the plain variable name.
```javascript
const vars = {
  backgroundColor: {
    value: '#0f0f0f',
    label: 'Background Color'
  }
}
```

Any other options you pass will be applied directly to the input, for example:
```javascript
const vars = {
  myVariable: {
    value: 'some text',
    foo: 'bar'
  }
}
```
will result in
```html
<input id='myVariable' type='text' value='some text' foo='bar'>
```

### Buttons

Setting `type: 'button'` will create a `<button>` element with the text of the label as its text. Let's look at how we can make the button do something when you press it using a callback.

### Callbacks

Each input can be passed a callback to run when the value is changed.
```javascript
const vars = {
  blendMode: {
    value: ['ADD', 'BLEND'],
    label: 'Blend Mode',
    callback: (val) => {
      console.log(val)
    }
  }
}
```

When creating your UI you can pass an additional `context` parameter which will be available to the callback. This can be useful if you need to access properties of other objects. For example, this will call a method on our `p5.js` instance.
```javascript
const vars = {
  loop: {
    value: true,
    callback: (val, p) => {
      val == true ? p.loop() : p.noLoop()
    }
  }
}

new p5(p => {
  p.setup = () => {
    const ui = new UI(variables, p).proxy
  }
})
```

You can also do something when a button is pressed. The button has no value but can still access the context if available.
```javascript
const vars = {
  myButton: {
      type: 'button',
      label: 'Click me',
      callback() {
        console.log('button clicked!')
      }
    }
  }
```

## To Do
- [ ] drag handle should be clearer and shouldn't scroll away
- [ ] make clearing saved state a function that can be called externally not hard-coded in
- [ ] document with gifs
- [ ] document multiple instances and new options
- [ ] fieldsets
- [ ] use with script tag only
- [ ] worked examples
- [ ] theming/styling
- [ ] transpile/minify?
- [ ] less flaky jsdom setup
- [ ] show/hide options?
- [ ] can callbacks reference the parent UI element?
- [ ] url encoding?
- [ ] load/save presets?


changelog.md
# Changelog

## [Unreleased]
### Fixed
- Remove incorrect window debug assignment

## [0.0.5] - 2019-01-05
### Added
- CSSLint
- Drag to move and resize UI panel
- Save and load values, position and size to localStorage automatically

### Changed
- Space out buttons
- Selects now need an options array, separate from value
- Show/hide with h key instead of button

## [0.0.4] - 2018-12-02
### Changed
- Support multiple UI instances
- Allow panel to scroll if taller than window

## [0.0.3] - 2018-11-15
### Fixed
- Edge fixes

### Changed
- update monospace font stack

### Added
- show/hide
- button support
- custom labels

### Removed
- removed global styling
- remove style/css loaders. End user will have to bring their own.

## [0.0.2] - 2018-11-11
### Fixed
- Correct main field in package.json

### Changed
- Start documentation

## [0.0.1] - 2018-11-10
### Changed
- Initial release


test/setup.js
import chai from 'chai'
import chaiDom from 'chai-dom'
global.expect = chai.use(chaiDom).expect


test/ui.test.js
import UI from '@/ui.js'

context('ui.js', () => {
  let ui, ui2

  beforeEach(() => {
    ui = new UI(
      {testVar: {value: 127}},
      {uid: 'id', save: false}
    )
  })

  afterEach(() => {
    if(ui) ui = ui.destroy()
  })

  describe('parses options', () => {
    it('sets the selector to body', () => {
      expect(ui.options.selector).to.equal('body')
    })

    it('sets the uid to id', () => {
      expect(ui.options.uid).to.equal('id')
    })
  })

  describe('creates basic UI elements', () => {
    it('creates the panel', () => {
      expect(document.querySelector('.sketch-ui-panel')).to.be.ok
    })

    it('creates the input', () => {
      const input = document.getElementById('testVar-id')
      expect(input.value).to.eq('127')
    })

    describe('creates a label', () => {
      it("for attribute matches the input's ID", () => {
        const label = document.querySelector('[for="testVar-id"]')
        expect(label).to.be.ok
      })

      it("label's text matches variable name", () => {
        const label = document.querySelector('[for="testVar-id"]')
        expect(label).to.contain.text('testVar')
      })

      describe('label', () => {
        beforeEach(() => {
          ui.destroy()
          ui = new UI(
            {testVar: {value: 127, label: 'My test variable'}},
            {uid: 'id'})
        })

        it('can be custom', () => {
          const label = document.querySelector('[for="testVar-id"]')
          expect(label).to.contain.text('My test variable')
        })
      })

      it('adds a span with the variable value', () => {
        const span = document.querySelector('[for="testVar-id"] span')
        expect(span).to.have.text('127')
      })
    })

    it('can be hidden', () => {
      const event = new Event('keyup')
      event.key = 'h'
      document.dispatchEvent(event)
      const panel = document.querySelector('.sketch-ui-panel')
      expect(panel).to.have.class('hidden')
    })

    it('cleans up after itself', () => {
      if(ui) ui = ui.destroy()
      const result = document.body.innerHTML.replace(/\s+/g, '')
      expect(result).to.equal('')
    })
  })

  describe('creates selects', () => {
    beforeEach(() => {
      ui = new UI(
        {testArray: {options: ['one', 'two']}},
        {uid: 'id'}
      )
    })

    it('creates the select', () => {
      expect(document.querySelector('select#testArray-id')).to.be.ok
    })

    it('creates options with the right value', () => {
      expect(document.querySelector('option').value).to.equal('one')
    })

    it('creates the right number of options', () => {
      expect(document.querySelectorAll('option').length).to.equal(2)
    })
  })

  describe('getting and setting via proxy works', () => {
    let proxy

    beforeEach(() => {proxy = ui.proxy})

    it('can get a value', () => {
      expect(proxy.testVar).to.equal(127)
    })

    it('can set a value', () => {
      proxy.testVar = 200
      expect(proxy.testVar).to.equal(200)
    })

    it('updates the field', () => {
      proxy.testVar = 200
      const input = document.querySelector('#testVar-id')
      expect(input.value).to.equal('200')
    })

    it('updates the span', () => {
      proxy.testVar = 200
      const span = document.querySelector('[for="testVar-id"] span')
      expect(span).to.have.text('200')
    })

    describe('respects limits', () => {
      it("won't set < min", () => {
        proxy.testVar = -1000
        expect(proxy.testVar).to.equal(0)
      })

      it("won't set > max", () => {
        proxy.testVar = 1000
        expect(proxy.testVar).to.equal(255)
      })

      it('returns undefined for a missing variable', () => {
        expect(proxy.foo).to.be.undefined
      })
    })
  })

  describe('passing a callback works', () => {
    let context

    beforeEach(() => {
      context = {called: false}

      ui = new UI({testVar: {
        value: 127,
        callback: function(val, context) {
          context.called = true
        }
      }}, {context: context})
    })

    it('calls the callback', () => {
      ui.proxy.testVar = 200
      expect(context.called).to.be.true
    })
  })

  describe('creates buttons', () => {
    let context

    beforeEach(() => {
      context = {called: false}
      ui = new UI({button: {
        type: 'button',
        callback: function(context) {
          context.called = true
        }
      }}, {context: context, uid: 'id'})
    })

    it('calls the callback when clicked', () => {
      const evt = new Event('click')
      const button = document.getElementById('button-id')
      button.dispatchEvent(evt)
      expect(context.called).to.be.true
    })
  })

  describe('getValues', () => {
    it('serialises values to an array of objects', () => {
      expect(ui.getValues()).to.deep.equal([{testVar: 127}])
    })
  })

  describe('getState', () => {
    it('returns an object with the dimensions and values', () => {
      expect(ui.getState()).to.deep.include({
        // why don't width/height work?
        // width: 320,
        // height: 450,
        xpos: 0,
        ypos: 0,
        values: [{testVar: 127}]
      })
    })
  })

  describe('can create multiple UIs', () => {
    beforeEach(() => {
      if(!document.querySelector('.ui-2-wrapper')) {
        const div = document.createElement('div')
        div.classList.add('ui-2-wrapper')
        document.body.appendChild(div)
      }
      ui2 = new UI(
        {testVar: {value: 127}},
        {uid: 'id2', save: false, selector: '.ui-2-wrapper'}
      )
    })

    afterEach(() => {
      if(ui2) ui2 = ui2.destroy()
    })

    it('creates two panels', () => {
      expect(document.querySelectorAll('.sketch-ui-panel').length).to.eq(2)
    })

    describe('they have separate values', () => {
      it('internally', () => {
        ui2.proxy.testVar = 12
        expect(ui.proxy.testVar).to.eq(127)
      })

      it('in the dom', () => {
        ui2.proxy.testVar = 12
        const result1 = document.querySelector('#testVar-id').value == '127'
        const result2 = document.querySelector('#testVar-id2').value == '12'
        expect(result1).to.be.true
        expect(result2).to.be.true
      })
    })
  })


})


test/.eslintrc.js
module.exports = {
  env: {
    mocha: true
  },
  globals: {
    expect: false
  }
}


src/index.js
import p5 from 'p5'
import './index.css'
import UI from './ui.js'

let state = {
  strokeColor: {
    value: '#ff0000'
  },
  strokeAlpha: {
    value: 100
  },
  backgroundColor: {
    value: '#202020'
  },
  diameter: {
    value: 50,
    max: 500
  },
  step: {
    value: 5,
  },
  animate: {
    value: false,
    callback: function(val, p) {
      val == true ? p.loop() : p.noLoop()
    }
  },
  option: {
    value: ['one', 'two', 'three']
  }
}

new p5(p => {
  let ui, v

  p.setup = () => {
    ui = new UI(state, p)
    v = ui.proxy
    p.createCanvas(p.windowWidth, p.windowHeight)
    p.background(v.backgroundColor)
    p.noLoop()
  }

  p.windowResized = () => {
    p.resizeCanvas(p.windowWidth, p.windowHeight)
    p.background(v.backgroundColor)
  }

  p.draw = () => {
    p.noFill()
    const str = p.color(v.strokeColor)
    str.setAlpha(v.strokeAlpha)
    p.stroke(str)
    p.ellipse(p.windowWidth/2, p.windowHeight/2, v.diameter)

    const newDiameter = v.diameter + p.map(p.noise(Date.now()), 0, 1, v.step * -0.5, v.step * 0.5)
    v.diameter = newDiameter
  }

  p.doubleClicked = () => {
    v.diameter, 250
    p.background(v.backgroundColor)
  }
})


src/utils.js
function validate(obj, prop, value) {
  if(obj[prop].min !== undefined) value = Math.max(value, obj[prop].min)
  if(obj[prop].max !== undefined) value = Math.min(value, obj[prop].max)
  return value
}

function inferType(value) {
  if(typeof value == 'string' && value[0] == '#' && value.length == 7) return 'color'
  if(typeof value == 'number') return 'range'
  if(typeof value == 'boolean') return 'checkbox'
  else return 'text'
}

function defaultSettings(type) {
  const defaults = {
    range: {
      step: 1,
      min: 0,
      max: 255
    }
  }

  return defaults[type]  ? defaults[type] : {}
}

function panelStyle(width, height, transform = {x: 0, y: 0}) {
  return `width: ${width}px; height: ${height}px; transform: translateX(${transform.x}px) translateY(${transform.y}px);`
}

export { validate, inferType, defaultSettings, panelStyle }

webpack.common.js
const path = require('path')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const CleanWebpackPlugin = require('clean-webpack-plugin')

module.exports = {
  entry: {
    app: './src/index.js'
  },
  resolve: {
    alias: {
      '@': path.resolve(__dirname, 'src')
    }
  },
  plugins: [
    new CleanWebpackPlugin(['dist']),
    new HtmlWebpackPlugin({
      title: 'Webpack Test',
      template: 'index.html'
    })
  ],
  output: {
    filename: '[name].[contenthash].js',
    path: path.resolve(__dirname, 'dist')
  },
  optimization: {
    splitChunks: {
      chunks: 'all'
    }
  },
  module: {
    rules: [
      {
        test: /\.css$/,
        use: [
          //TODO: use mini-css-extract-plugin in production
          'style-loader',
          'css-loader'
        ]
      }
    ]
  }
}


src/ui.js
import { validate, inferType, defaultSettings, panelStyle } from './utils.js'

export default class UI {
  constructor(variables, options) {
    this.variables = variables
    const defaultOptions = {
      uid: 'ui-' + Math.floor(Math.random() * Date.now()),
      selector: 'body',
      save: false,
      width: 320,
      height: 450,
      xpos: 0,
      ypos: 0
    }
    if(!options) options = {}
    this.options = Object.assign(defaultOptions, options)

    if(this.options.save) this.readSavedState()

    this.createUI(this.variables)
    this.revocable = this.createProxy()
    this.proxy = this.revocable.proxy
  }

  destroy() {
    this.panel.remove()
    this.revocable.revoke()
    return null
  }

  readSavedState() {
    const savedState = window.localStorage.getItem(this.options.uid)
    if(savedState) {
      const parsedState = JSON.parse(savedState)

      parsedState.values.forEach(val => {
        const name = Object.keys(val)[0]
        const value = Object.values(val)[0]
        if(this.variables[name]) this.variables[name].value = value
      })

      delete parsedState.values
      this.options = Object.assign(this.options, parsedState)
    }
  }

  getState() {
    return {
      width: this.options.width,
      height: this.options.height,
      xpos: this.options.xpos,
      ypos: this.options.ypos,
      values: this.getValues()
    }
  }

  saveState() {
    if(this.options.save) {
      const newState = JSON.stringify(this.getState())
      window.localStorage.setItem(this.options.uid, newState)
    }
  }

  getValues() {
    let values = []
    for(let v in this.variables) {
      if(this.variables[v].type != 'button') {
        values.push({[v]: this.variables[v].value})
      }
    }
    return values
  }

  createProxy() {
    return Proxy.revocable(this.variables, {
      get: (obj, prop) => {
        if (obj[prop]) return obj[prop].value
        else {
          // eslint-disable-next-line no-console
          console.warn(`No variable ${prop} found`)
          return undefined
        }
      },
      set: (obj, prop, value) => {
        value = validate(obj, prop, value)
        obj[prop].value = value
        this.updateField(prop, value)
        if(obj[prop].callback) obj[prop].callback(value, this.options.context)
        this.saveState()
        return true
      }
    })
  }

  updateSize() {
    const opts = this.options
    opts.width = this.panel.offsetWidth
    opts.height = this.panel.offsetHeight
    const newStyle = panelStyle(opts.width, opts.height, {x: opts.xpos, y: opts.ypos})
    this.panel.setAttribute('style', newStyle)
    this.saveState()
  }

  updateField(prop, value) {
    const type = this.variables[prop].type
    const el = document.getElementById(this.attrWithUid(prop))
    if(type === 'checkbox') {
      value === true ? el.checked = true : el.removeAttribute('checked')
    } else el.value = value

    document.querySelector(`[for=${this.attrWithUid(prop)}] span`).innerHTML = value
  }

  createUI(object) {
    const panel = this.buildPanel()

    for(let item in object) {
      let attrs = object[item]

      if(!attrs.type) {
        if(attrs.options) {
          attrs.type = 'select'
        } else {
          attrs.type = inferType(attrs.value)
        }
      }

      const defaults = defaultSettings(attrs.type)
      // Spread with object literals not supported in Edge
      // object [item] = {...defaults, ...attrs}
      object[item] = Object.assign(defaults, attrs)

      if(attrs.type != 'button') {
        panel.appendChild(this.buildLabel(item, object[item].label))
      }
      panel.appendChild(this.buildInput(item, object[item]))
    }

    this.panel = panel
    this.updateSize()
    return panel
  }

  buildPanel(selector = this.options.selector) {
    const el = document.querySelector(selector)
    el.innerHTML = `
    <div class='sketch-ui-panel' id='${this.options.uid}' style='${panelStyle(this.options.width, this.options.height)}'>
      <div class='sketch-ui-handle'>Move</div>
      <button class='ui-clear'>Clear saved</button>
    </div>
  `
    const panel = el.querySelector(`#${this.options.uid}`)
    document.addEventListener('keyup', event => {
      if(event.key == 'h') panel.classList.toggle('hidden')
    })

    panel.querySelector('.ui-clear')
      .addEventListener('click', () => window.localStorage.clear())

    const handle = panel.querySelector('.sketch-ui-handle')
    handle.addEventListener('mousedown', event => {
      event.preventDefault()
      const rect = panel.getBoundingClientRect()
      const scroll = document.querySelector(this.options.selector).getBoundingClientRect().top
      const opts = this.options
      opts.xpos = rect.left
      opts.ypos = rect.top - scroll
      opts.xOffset = event.clientX - opts.xpos
      opts.yOffset = event.clientY - opts.ypos

      const callback = this.handlePanelDrag.bind(this)
      document.addEventListener('mousemove', callback)
      document.addEventListener('mouseup', event => {
        event.preventDefault()
        document.removeEventListener('mousemove', callback)
      })
    })

    panel.addEventListener('mouseup', () => this.updateSize())

    return panel
  }

  handlePanelDrag(event) {
    this.options.xpos = event.clientX - this.options.xOffset
    this.options.ypos = event.clientY - this.options.yOffset
    this.updateSize()
  }

  buildLabel(name, text = name) {
    const label = document.createElement('label')
    label.setAttribute('for', this.attrWithUid(name))
    label.innerHTML = `${text}<span></span>`
    return label
  }

  buildInput(name, variables) {
    if(variables.type === 'button') return this.buildButton(name, variables)
    if(variables.type === 'select') return this.buildSelect(name, variables)

    let input = document.createElement('input')
    input.setAttribute('id', this.attrWithUid(name))

    document.querySelector(`[for=${this.attrWithUid(name)}] span`).innerHTML = variables.value

    const exclusions = ['callback', 'label', 'value']

    for(let attr in variables) {
      if(!exclusions.includes(attr)) {
        input.setAttribute(attr, variables[attr])
      }
    }

    input.value = variables['value']

    if(variables.type == 'checkbox') {
      input.removeAttribute('value')
      if(variables.value) input.setAttribute('checked', variables.value)

      input.addEventListener('change', event => {
        this.proxy[name] = event.target.checked
      })
    }

    else {
      input.addEventListener('input', event => {
        this.proxy[name] = event.target.value
      })
    }

    return input
  }

  buildButton(name, variables) {
    const div = document.createElement('div')
    div.classList.add('sketch-ui-button-wrapper')
    div.innerHTML = `
      <button id=${this.attrWithUid(name)}>
        ${variables.label ? variables.label : name}
      </button>
    `

    if(variables.callback) {
      const button = div.querySelector('button')
      button.addEventListener('click', event => {
        event.preventDefault()
        variables.callback(this.options.context)
      })
    }

    return div
  }

  buildSelect(name, variables) {
    let select = document.createElement('select')
    select.setAttribute('id', this.attrWithUid(name))
    const initialValue = variables.value ? variables.value : variables.options[0]
    variables.value = initialValue
    select.innerHTML = variables.options.map(option => {
      const selected = option == initialValue ? 'selected' : ''
      return `
        <option ${selected} value='${option}'>
          ${option}
        </option>
      `}).join('')

    document.querySelector(`[for=${this.attrWithUid(name)}] span`).innerHTML = initialValue
    select.addEventListener('change', event => this.proxy[name] = event.target.value)
    return select
  }

  attrWithUid(attr) { return `${attr}-${this.options.uid}` }
}
